lib/sgpres.gi

```diff
@@ -2766,38 +2766,1163 @@ end);
+
+  for w in [1..Length(subgens)] do
+    if DATA.augmented then
+      NEWTC_ModifiedScanAndFill(DATA,1,DATA.subgword[w],[w]);
+    else
+      NEWTC_ScanAndFill(DATA,1,DATA.subgword[w]);
+    fi;
+  od;
+  NEWTC_ProcessDeductions(DATA);
+
+  # words we want to trace early (as they might reduce the number of
+  # definitions
+  if trace<>false then
+    #trace:=Concatenation(trace,ri); #don't seem to help
+    for w in trace do
+      if IsList(w[1]) then
+	w:=w[1]; # get word from value
+      fi;
+      i:=1;
+      for a in w do
+	if ct[a+offset][i]=0 then
+	  NEWTC_Define(DATA,i,a);
+	  NEWTC_ProcessDeductions(DATA);
+	  i:=p[i]; # in case there is a change
+	fi;
+	i:=ct[a+offset][i];
+      od;
+    od;
+  fi;
+
+  i:=1;
+  while i<=DATA.n do
+    for a in A do
+      if p[i]=i then
+	if ct[a+offset][i]=0 then
+	  NEWTC_Define(DATA,i,a);
+	  oldead:=DATA.dead;
+	  NEWTC_ProcessDeductions(DATA);
+	  if PERCFACT*(DATA.dead-oldead)>DATA.n then
+	    if DATA.n>1000 then
+	      Info( InfoFpGroup, 2, "\t", DATA.defcount, "\t\t", DATA.dead,
+	      "\t\t", DATA.n-DATA.dead, "\t\t", DATA.n );
+	    fi;
+	    if IsBound(DATA.with) then
+	      # collapse -- find collapse word
+	      # in two different ways (as they can differ after compression)
+
+	      # first trace through the coset table, this uses the prior
+	      # reductions
+	      j:=i;
+	      while j<>p[j] do
+		j:=p[j];
+	      od;
+	      w:=[a]; # last letter added
+	      while j<>1 do
+		Assert(2,j=p[j]);
+		Add(w,with[j]);
+		Assert(2,0<>ct[-with[j]+offset][j]);
+		j:=ct[-with[j]+offset][j]; # unapply this generator
+	      od;
+
+	      # free reduce -- partial collapse can lead to not free cancellation
+	      # and fix order
+	      w:=Reversed(FreelyReducedLetterRepWord(w));
+	      #w1:=w;
+
+	      j:=PositionProperty(collapse,x->x[1]=w);
+	      if j=fail then
+		Add(collapse,[w,DATA.dead-oldead]); # word that caused a collapse
+	      else
+		collapse[j][2]:=Maximum(collapse[j][2],DATA.dead-oldead);
+	      fi;
+
+	      # now use the `from' list (which does not collapse under
+	      # coincidences, only under compression) and not the coset table,
+	      #  as it # keeps the old definition order, not yet using coincidence
+	      j:=i;
+	      w:=[a]; # last letter added
+	      while j<>1 do
+		Add(w,with[j]);
+		j:=from[j];
+	      od;
+
+	      # free reduce -- partial collapse can lead to not free
+	      # cancellation and fix order
+	      w:=Reversed(FreelyReducedLetterRepWord(w));
+
+	      j:=PositionProperty(collapse,x->x[1]=w);
+	      if j=fail then
+		Add(collapse,[w,DATA.dead-oldead]); # word caused collapse
+	      else
+		collapse[j][2]:=Maximum(collapse[j][2],DATA.dead-oldead);
+	      fi;
+
+	      Info(InfoFpGroup,3,"collapse ",DATA.dead-oldead);
+
+	    fi;
+	  fi;
+
+	fi;
+      fi;
+    od;
+
+    # at least 33% trash (3=1+1/0.3, 11=1+1/0.1), or over limtrigger and
+    # at least 2% trash
+    if
```

Comment:
I don't quite understand this comment, as it's missing a verb. I guess it is meant to say something like "If there is more than this amount of trash, compress the table" ? It would be nice if the comment would state that more clearly.